src/net_processing.cpp

```diff
@@ -334,7 +334,10 @@ class PeerManagerImpl final : public PeerManager
 
     /** Send a ping message every PING_INTERVAL or if requested via RPC. May
      *  mark the peer to be disconnected if a ping has timed out. */
-    void MaybeSendPing(CNode& node_to, Peer& peer);
+    void MaybeSendPing(CNode& node_to, Peer& peer, std::chrono::microseconds now);
```

Comment:
looks like both `MaybeSendPing` and `MaybeSendAddr` can be made into const member functions

Reply:
I'm not sure that `const` should be used in cases like this. The compiler will make sure that any function marked `const` doesn't mutate the object's data members, which is the case for both of these functions. However, `const` is not transitive, so the functions are still able to modify state through pointers and modify global state.

Here, `PeerManagerImpl` owns the map of `Peer` objects:

```
class PeerManagerImpl final : public PeerManager
{
[...]
private:
    /**
     * Map of all Peer objects, keyed by peer id. This map is protected
     * by the m_peer_mutex. Once a shared pointer reference is
     * taken, the lock may be released. Individual fields are protected by
     * their own locks.
     */
    std::map<NodeId, PeerRef> m_peer_map GUARDED_BY(m_peer_mutex);
```

and the map of `CNodeState` objects are global (in the file's unnamed namespace):

```
/** Map maintaining per-node state. */
static std::map<NodeId, CNodeState> mapNodeState GUARDED_BY(cs_main);
```

So `PeerManagerImpl` definitely owns the `Peer` objects, and logically owns the `CNodeState` objects since it's responsible for constructing and destructing them (and #20758 actually moves the `CNodeState` objects into `PeerManagerImpl`).

These functions don't mutate `PeerManagerImpl`, but _do_ mutate data that it owns. I think in such cases there's an argument that we shouldn't mark the function `const` even if the compiler would allow us to. For me, `const` on a member function is telling the programmer "this function does not mutate the logical state of this object", which isn't the case here.

Reply:
Since `PeerRef` is a shared_ptr which does its own destruction once the ref count hits zero, I don't think it's accurate to say that `PeerManagerImpl` owns the `Peer` object.

Reply:
I'm not sure it's right to say that ownership of Peer is shared because it uses a shared pointer. We use a shared pointer for the Peer for convenience/safety/synchronization. We never take a shared pointer to a Peer outside PeerManagerImpl, and if we really wanted to, could implement it with raw pointers and a manual ref count, and then `new` and `delete` the Peer objects (as is done by CConnman for its CNode objects in vNodes)

That's maybe a bit pedantic. I think the more important question is "what is `const` for?" For me, it means "this function won't mutate the _logical_ state of this object". If we're using it to say "this function won't mutate the data members of this object, but it will mutate the state of data structures that are integral to how this object operates, but will do so through a pointer", then it seems a bit legalistic and not very useful to me.

Reply:
I was surprised to discover that if you have a class Y with a member `X& x` then C++ thinks it's fine to modify `x.foo = 3` from a const member function of Y. So not sure I'd argue against "a bit legalistic and not very useful"... (So in particular, if you have a `const PeerManagerImpl`, it will have no problems calling non-const functions of its `m_connman`)

If you consider modifying the contents of a `PeerRef` to be modifying a `PeerManagerImpl`, then I don't think it's consistent to mark `GetPeerRef()` as const either (code compiles fine that way though). I dunno; "shared" makes me think things are shared, not owned, so that's definitely a mental adjustment.

I guess I don't think it's a good idea to try to put too much convention on top of what our compilers will actually enforce here though.

Reply:
This certainly isn't a hill I'm prepared to die on (either here or in https://github.com/bitcoin/bitcoin/pull/21162#discussion_r587410937). My preference is not to mark functions as const if they're mutating data that is so closely associated with the object that it can only be modified through the object, but I'll go along with the majority if people want this to be const.

Reply:
Yeah, not my hill either, but I think it's worth discussing so we've got a *shared* understanding of how this data's meant to be controlled... I have some other ideas about `Peer` ownership I want to try out once more of this refactoring is finished that might change this, and worst case can add or remove const markers when we're reordering the PeerManageImpl declarations and adding more comments...

Reply:
I agree that the relevant question is "what is `const` for?" The two main aspects are compiler enforcement & human understanding. @jnewbery I get what you're saying about making these member functions `const` is misleading for the humans, because the function often mutates the logical state of the object. However, I also agree with @ajtowns that since the compilers would actually be enforcing something here, it does offer a bit of literal guarding?

I don't have a strong opinion on which way we go, but it is definitely an interesting consideration & mostly I conclude that `const` is complex. I'm personally curious to learn more about how a compiler uses the keyword- when its just enforcement vs optimization. (recommending resources welcome!)

Reply:
I'm 99% sure that compilers don't use `const` markers on functions/pointers/references at all for optimization. All they do is disallow certain operations on them.

It's different if you have an actual `const` variable (e.g. a global, local, or member variable that is declared const where it is defined & initialized), because that means "changing this is UB". Any other use simply means "no non-const operations are allowed on object *through this reference/pointer/function*; if other code is invoked that happens to hold a non-const reference/pointer, the object can still change (and the compiler can't assume a cached version in a register won't change e.g.).

Reply:
@sipa yes, that's exactly my understanding. Mutating a `const`-declared variable is UB. Therefore compilers could make optimizations based on the fact that `const`-declared variables cannot change.

`const` on a member functions simply means that this function call won't mutate the state of the object, although of course, other code could mutate the state. Therefore, the compiler can't make any assumptions about the state of the variable not changing.

You can even mutate the state through a `const` function by casting away the constness, eg:

```c++
#include <iostream>

class thing
{
    int x{0};

public:
    const void mutate(int x_in)
    {
        thing* mutable_this = const_cast<thing*>(this);
        mutable_this->x = x_in;
    }

    const int access() { return x; }

    thing(int x_in) : x(x_in) { }
};
 
int main()
{
    thing t(1);
    std::cout << "t.x = " << t.access() << "\n";

    t.mutate(2);
    std::cout << "t.x = " << t.access() << "\n";

    return 0;
}
```

You can see that the state of `t` has been mutated in the `mutate()` function, even though that function is `const`. This compiles with no warnings. I believe this is perfectly legal (although ill-advised!) c++. The c++ core guidelines advise against casting away const in almost all cases: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es50-dont-cast-away-const. Scott Meyers does use casting away const to avoid code duplication: https://stackoverflow.com/a/123995/933705 (although that book is now quite old, so I'm not sure if the advice is still current).

Going back to the main question about whether these functions _should_ be `const`, I found [the isocpp.org wikipedia page on const correctness](https://isocpp.org/wiki/faq/const-correctness) to be very useful in clarifying my thoughts. A few extracts:

> The trailing `const` on `inspect()` member function should be used to mean the method won’t change the object’s _abstract_ (client-visible) state. That is slightly different from saying the method won’t change the “raw bits” of the object’s `struct`

> a `const` member function must not change (or allow a caller to change) the this object’s _logical_ state (AKA _abstract_ state AKA _meaningwise_ state). Think of what an object means, not how it is internally implemented. A Person’s age and name are logically part of the Person, but the Person’s neighbor and employer are not. An inspector method that returns part of the this object’s _logical_ / _abstract_ / _meaningwise_ state must not return a non-`const` pointer (or reference) to that part, **independent of whether that part is internally implemented as a direct data-member physically embedded within the this object or some other way**. (emphasis mine)

> - If a method changes any part of the object’s logical state, it logically is a mutator; it should not be const even if (as actually happens!) the method doesn’t change any physical bits of the object’s concrete state.
> - Conversely, a method is logically an inspector and should be const if it never changes any part of the object’s logical state, even if (as actually happens!) the method changes physical bits of the object’s concrete state.

> If you are outside the class — you are a normal user, every experiment you could perform (every method or sequence of methods you call) would have the same results (same return values, same exceptions or lack of exceptions) irrespective of whether you first called that lookup method. If the lookup function changed any future behavior of any future method (not just making it faster but changed the outcome, changed the return value, changed the exception), then the lookup method changed the object’s logical state — it is a mutuator.

Based on that, I believe these functions should not be marked `const`. Calling `MaybeSendPing()` and `MaybeSendAddr()` multiple times will result in different behaviour that is observable from outside the class, since the `Peer()` objects are mutated.

Reply:
I'm convinced :)